### PR TITLE
CLI: Add UUID as attribute and make searchable

### LINKED ISSUE #4112: CLI: display the UUID in the show command
## Description

Add Uuid as a default attribute to list in `show` command or as an attribute with `-a option` to be able to retrive it

`keepassxc-cli show path/to/database 'search'`

Should show:

```
Insert password to unlock path.kdbx: 
Title: mi title
UserName: username
Password: *****
URL: 
Uuid: 9340590348503485034
Notes: 
```

`keepassxc-cli show -a Uuid path/to/database 'search'`

Should show:

```
Insert password to unlock path.kdbx: 
9340590348503485034
```

## Context
It helps to integrate with other applications because it guaranties no duplicate entries.

### LINKED ISSUE #4113: keepassxc-cli 2.5.2 show stoped finding by using Uuid
# Summary

`keepassxc-cli show database.kdbx '1706a7eea30d5c49b283645524006371'`

Should locate entry by UUID, version 2.3.1, allow to find by Uuid

## Debug Info
keepassxc-cli , Version 2.5.2

Operating system: Linux, Mint
CPU architecture: 686

Enabled extensions:
- None

### PR DESCRIPTION
**Pull request by bahelit**
Add UUID as an attribute when using keepassxc-cli show. Also makes this a searchable attribute.

Fixes #4112 and fixes #4113.

Testing is validated by running the command and searching by the UUID and seeing UUID in the output:
`keepassxc-cli show ~/Documents/PasswordsTest.kdbx '076b5698-5ad7-4c47-97a7-c5e373efa0b3'`

> Title: Cool Site
> UserName: derp
> Password: PROTECTED
> URL: punny.xyz
> Uuid: 076b5698-5ad7-4c47-97a7-c5e373efa0b3
> Notes:

### CODE REVIEW

File: src/cli/Show.cpp [hunk 1]
```diff
@@ -96,6 +96,8 @@ int Show::executeWithDatabase(QSharedPointer<Database> database, QSharedPointer<
         }
         if (entry->attributes()->isProtected(attributeName) && showAttributeNames && !showProtectedAttributes) {
             outputTextStream << "PROTECTED" << endl;
+        } else if (attributeName.compare("Uuid") == 0) {
+            outputTextStream << entry->uuid().toString().remove((QRegExp("[{}]"))) << endl;
```
Comment (phoerious) [inline src/cli/Show.cpp]: Use `entry->uuid().toString(QUuid::WithoutBraces)` (`QRegExp` is deprecated BTW).

File: src/core/Group.cpp [hunk 2]
```diff
@@ -677,6 +677,11 @@ Entry* Group::findEntryByPathRecursive(const QString& entryPath, const QString&
             || (!entryPath.startsWith("/") && entry->title() == entryPath)) {
             return entry;
         }
+        bool found = false;
+        found = entry->uuid() == QUuid::fromRfc4122(QByteArray::fromHex(entry->uuid().toString().toLatin1()));
+        if (found) {
+            return entry;
+        }
         // clang-format on
```
Comment (phoerious) [inline src/core/Group.cpp]: I don't understand what you are trying to accomplish here. `entry->uuid() == QUuid::fromRfc4122(QByteArray::fromHex(entry->uuid().toString().toLatin1()))` is always true and an entirely unnecessary back-and-forth chain of conversions.
The loop now exits with an arbitrary entry unless you get very lucky and the entry you are actually looking for is the very first one in the list.

File: src/cli/Show.cpp (hunk 1 above)
Comment (bahelit) [reply src/cli/Show.cpp]: Got it, thanks.

File: src/core/Group.cpp (hunk 2 above)
Comment (bahelit) [reply src/core/Group.cpp]: Grabbed this code from another location and missed this.
Ran another test and see that it is grabbed the 1st entry. Should've caught this :man_facepalming: .

Am matching the entry string correctly now with QUuid::Id128, as this removes the braces and dashes. Tested with multiple entries.

**Comment by bahelit** [comment]
The last commit added the include for quuid.h to satisfy the failing Ubuntu build but it still can't find it. Am running Ubuntu and see the path it's being pulled from is the same path as QLocale which was recently added. Any ideas why the Ubuntu build is not seeing the include?

File: src/cli/Show.cpp [hunk 3]
```diff
@@ -28,6 +28,7 @@
 #include "core/Group.h"
 
 #include <QLocale>
+#include <quuid.h>
```
Comment (phoerious) [inline src/cli/Show.cpp]: This should be `#include <QUuid>`.

File: src/core/Group.cpp [hunk 4]
```diff
@@ -26,6 +26,7 @@
 #include "core/Tools.h"
 
 #include <QtConcurrent>
+#include <quuid.h>
```
Comment (phoerious) [inline src/core/Group.cpp]: Same here.

File: src/core/Group.cpp [hunk 5]
```diff
@@ -677,6 +678,9 @@ Entry* Group::findEntryByPathRecursive(const QString& entryPath, const QString&
             || (!entryPath.startsWith("/") && entry->title() == entryPath)) {
             return entry;
         }
+        if (entryPath == entry->uuid().toString(QUuid::Id128)) {
+            return entry;
+        }
```
Comment (phoerious) [inline src/core/Group.cpp]: Search by path is not the same as searching by UUID and getting a UUID match is quite unexpected. This should be pulled out into a separate method.

File: src/cli/Show.cpp (hunk 3 above)
Comment (bahelit) [reply src/cli/Show.cpp]: Thank you for pointing that out :+1:

File: src/core/Group.cpp (hunk 4 above)
Comment (bahelit) [reply src/core/Group.cpp]: Thank you for pointing that out :+1:

File: src/core/Group.cpp (hunk 5 above)
Comment (bahelit) [reply src/core/Group.cpp]: Scrolled up just a little bit and saw the method already exists.
Added it after the 1st attempt to find by path fails.

Comment (bahelit) [reply src/core/Group.cpp]: Created the separate method.

**Comment by bahelit** [comment]
Don't get why the Ubuntu build is still not picking up the QUuid include. Are there scripts available for developers to reproduce the CI pipeline locally, possibly with docker?

**Comment by xvallspl** [comment]
What's the status on this PR? I see that parts of it (for instance, Group::findEntryByUuid) are already in the codebase. Is this PR valid anymore, is it open as a reminder of things left to implement...?

**Comment by droidmonkey** [comment]
This just needs to be rebased and tested.

**Comment by bahelit** [comment]
Am still encountering an error in the TeamCity builds. The import QUuid is the cause, when attempting to use it as an import and access the WithoutBraces enum the build fails. Can see that the file is an external dependency found in qt5/QtCore on my Ubuntu 20.04 system. The build also works locally using the release-tool inside the docker container. Put the regexp filter back in place so the build would pass. Updated a couple of the tests, there's still one failing and am looking into it.

**Comment by droidmonkey** [comment]
The code changes are totally incorrect for this feature. I'm just going to close this and there will have to be a reattempt.

**Pull request closed without merging**